Logger: Implement FunctionOrigin option, to substitute logger name for function's

Geod24 opened the pull request:
```
This follows the well-established practice of having a `Logger` per module,
or more granular. Having a Logger per function would not be practical,
and expensive, and adding a field to `Event` would lead to a duplication
of `Layout` (as e.g. `LayoutDate` would need to account for it).

The most practical way for the user seems to be to provide the ability
to substitute the Logger name for the name of the function the calls
originate from, which this implements.
```

Comes with a bunch of fixes / refactoring as I tried a few approaches before settling on this one.
Will test in Agora.

Geod24 committed:
Trivial: Correct LayoutSimple DDOC indentation

Use 4 spaces, not 3.

Geod24 committed:
Trivial: Re-enable LayoutSimple test with LDC

It was probably disabled due to an ancient bug which doesn't manifest anymore.

Geod24 committed:
Trivial: Correct comment

Geod24 committed:
Logger: Implement FunctionOrigin option, to substitute logger name for function's

This follows the well-established practice of having a `Logger` per module,
or more granular. Having a Logger per function would not be practical,
and expensive, and adding a field to `Event` would lead to a duplication
of `Layout` (as e.g. `LayoutDate` would need to account for it).

The most practical way for the user seems to be to provide the ability
to substitute the Logger name for the name of the function the calls
originate from, which this implements.

Geod24 committed:
Logger: Extend options handling to use flags

This should make new options easier to add, and handling more consistent,
for example by allowing 'propagate' to apply to any option.

Geod24 committed:
Logger: Make Hierarchy non-templated

Hierarchy was made templated when the Logger was re-written using the Formatter.
Nowadays, this templating serves no purpose, and conflict with the OOP approach
that Logger was attempting to take.

Geod24 committed:
Split propagateLevel from propagateValue

propagateValue is used to set an arbitrary field in a Logger to an arbitrary value,
recursively. While convenient, it falls short as soon as the field has some logic
to it. While this can be worked around by exposing specific setters that handle
such logic, we want to move away from simple boolean fields towards
an enum of options, which would mean changing propagateValue - However, as it's
currently used for both the Level and the boolean collect_stats, we first need
to split those two usages.

Geod24 committed:
Logger: Expose methods to get/set options

Instead of having individual getters/setters,
we can unify it under a singe method.

Geod24 committed:
Logger: Fix names of logger used in unittest

The Loggers are usually instantiated using Hierarchy.inject,
which always appends a dot at the end.
However, when we instantiate them manually, we don't do that,
so the name used for the event is actually 'dumm', not 'dummy'.

mkykadir commented on source/dtext/log/Logger.d:
```diff
@@ -65,6 +65,12 @@ version (unittest)
 /// Convenience alias to available log level, see `ILogger.Level`
 public alias ILogger.Level Level;
 
+/// List of options that can be set on a `Logger`
```
Let's have a comment indicating that this enum's values will be used in bitwise operations

Geod24 replied:
"Flag of options" ?

mkykadir commented on source/dtext/log/ILogger.d:
```diff
@@ -34,6 +34,15 @@ version (unittest)
+        /// Use the `Logger` ancestors' `Appender` as well as the `Logger`'s own
+        Additive,
+        /// Count emitted event towards the global stats
+        CollectStats,
```
May be instead of comment above;

```suggestion
        CollectStats = 1 << 0,
```

would indicate the bitwise operations

Geod24 replied:
Thanks 🤦

mkykadir commented on source/dtext/log/ILogger.d:
```diff
@@ -41,6 +41,8 @@ interface ILogger
         /// Count emitted event towards the global stats
         CollectStats,
+        /// Use function names instead of logger name for `Event.name`
+        FunctionOrigin,
```
```suggestion
        FunctionOrigin = 1 << 1,
```

mkykadir reviewed (approved):
`12222ed6` commit has typo;
`currenly`

Geod24 committed:
Logger: Make collect_stats a LogOption member

Like we did for `additive`, this simplifies propagation
and make the code more extensible in the future.

Geod24 committed:
Trivial: Remove superfluous column after enum decl

Geod24 commented:
Fixed
